core/src/test/java/org/elasticsearch/action/support/master/TransportMasterNodeActionTests.java

```diff
@@ -0,0 +1,336 @@
+        }
+
+        @Override
+        protected Response newResponse() {
+            return new Response();
+        }
+
+        @Override
+        protected void masterOperation(Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
+            listener.onResponse(new Response()); // default implementation, overridden in specific tests
+        }
+
+        @Override
+        protected ClusterBlockException checkBlock(Request request, ClusterState state) {
+            return null; // default implementation, overridden in specific tests
+        }
+    }
+
+    public enum MasterOptions {
+        MASTER_EXISTS,
+        MASTER_DOES_NOT_EXIST,
+    }
+
+    public enum LocalNodeMasterOptions {
+        LOCAL_NODE_IS_MASTER,
+        LOCAL_NODE_IS_NOT_MASTER,
+    }
+
+    private ClusterState mockedClusterState(MasterOptions existsMaster, LocalNodeMasterOptions localNodeIsMaster) {
+        DiscoveryNodes.Builder discoBuilder = DiscoveryNodes.builder();
+        discoBuilder.put(localNode);
+        discoBuilder.put(remoteNode);
+
+        discoBuilder.localNodeId(localNode.id());
+
+        if (existsMaster == MASTER_EXISTS) {
+            discoBuilder.masterNodeId((localNodeIsMaster == LOCAL_NODE_IS_MASTER ? localNode : remoteNode).id());
+        }
+
+        IndexMetaData indexMetaData = IndexMetaData.builder("test-idx").settings(Settings.builder()
+                .put(SETTING_VERSION_CREATED, Version.CURRENT)
+                .put(SETTING_NUMBER_OF_SHARDS, 1).put(SETTING_NUMBER_OF_REPLICAS, 1)
+                .put(SETTING_CREATION_DATE, System.currentTimeMillis())).build();
+        ClusterState.Builder state = ClusterState.builder(new ClusterName("test"));
+        state.nodes(discoBuilder);
+        state.metaData(MetaData.builder().put(indexMetaData, false).generateClusterUuidIfNeeded());
+        return state.build();
+    }
+
+    @Test
+    public void testLocalOperationWithoutBlocks() throws ExecutionException, InterruptedException {
+        final boolean masterOperationFailure = randomBoolean();
+
+        Request request = new Request();
+        PlainActionFuture<Response> listener = new PlainActionFuture<>();
+
+        final Throwable exception = new Throwable();
+        final Response response = new Response();
+
+        clusterService.setState(mockedClusterState(MASTER_EXISTS, LOCAL_NODE_IS_MASTER));
+
+        new Action(Settings.EMPTY, "testAction", transportService, clusterService, threadPool) {
+            @Override
+            protected void masterOperation(Request request, ClusterState state, ActionListener<Response> listener) throws Exception {
+                if (masterOperationFailure) {
+                    listener.onFailure(exception);
+                } else {
+                    listener.onResponse(response);
+                }
+            }
+        }.execute(request, listener);
+
+        if (masterOperationFailure) {
+            try {
```

Comment:
can we add an assertTrue(listener.isDone())? we expect the operation to return immediately I think?